src/main/network/router/token_bucket.rs

```diff
@@ -0,0 +1,154 @@
+use std::time::Duration;
+
+use crate::core::support::emulated_time::EmulatedTime;
+use crate::core::support::simulation_time::SimulationTime;
+use crate::core::worker::Worker;
+
+pub struct TokenBucket {
+    capacity: u64,
+    balance: u64,
+    refill_size: u64,
+    refill_interval: Duration,
+    last_refill: EmulatedTime,
+}
+
+impl TokenBucket {
+    /// Creates a new token bucket rate limiter. The capacity enables
+    /// burstiness, while the long term rate is defined by refill_size tokens
+    /// being periodically added to the bucket every refill_interval duration.
+    /// Returns None if any of the args are non-positive.
+    pub fn new(capacity: u64, refill_size: u64, refill_interval: Duration) -> Option<TokenBucket> {
+        if capacity > 0 && refill_size > 0 && refill_interval.as_nanos() > 0 {
+            Some(TokenBucket {
+                capacity,
+                balance: capacity,
+                refill_size,
+                refill_interval,
+                last_refill: Worker::current_time().unwrap(),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Remove size tokens from the bucket if and only if the bucket contains
+    /// at least size tokens. Returns the updated token balance on success, and
+    /// the duration until the next token refill on error.
+    pub fn comforming_remove(&mut self, size: u64) -> Result<u64, Duration> {
+        let next_refill_span = self.lazy_refill();
+
+        if size <= self.balance {
+            self.balance -= size;
+            Ok(self.balance)
+        } else {
+            Err(next_refill_span)
+        }
```

Comment:
nit: maybe a little more clear as the following, then you don't need to worry about getting the condition and underflow stuff correct:

```rust
self.balance = self.balance.checked_sub(size).ok_or(next_refill_span)?;
Ok(self.balance)
```

Reply:
Done.